### DISCUSSION TITLE
How to modify threadblock shape when A/B is transposed in GEMM?

### DISCUSSION THREAD

**Discussion by imoneoi** [NVIDIA/cutlass#1281] [Q&A] [answered]
I wrote a custom kernel for grouped GEMM using cutlass. Should I transpose the threadblock, warp and instruction shapes [here](https://github.com/imoneoi/cutlass_grouped_gemm/blob/main/csrc/grouped_gemm.cu#L100-L102) accordingly when A/B is transposed?

Besides, how to tune these shapes for optimal performance?

**Comment by jackkosaian** [comment]
Do you mean when there is an "internal transposition" [due to the output being column major](https://github.com/NVIDIA/cutlass/blob/b7508e337938137a699e486d8997646980acfc58/include/cutlass/gemm/kernel/default_gemm_grouped.h#L198)? If so, you should not need to change those template parameters.

To tune for performance, it's suggested to try out many valid combinations of threadblock shape, warp shape, and stage count. You may find it useful to follow some of the combinations listed by the CUTLASS kernel generator. For example, [these lines](https://github.com/NVIDIA/cutlass/blob/b7508e337938137a699e486d8997646980acfc58/python/cutlass_library/generator.py#L1984-L2005) show a subset of vali combinations threadblock shape, stage count, and warp count for an SM80 FP16 GEMM.

**Reply by imoneoi** [reply]
Thank you for your response! What I mean is when the input matrix A or B is transposed (column major) while C is always row major. Also, if any input becomes column major, does the best performing threadblock/warp shape/stage count change?

**Reply by jackkosaian** [reply, marked as answer]
Thanks for clarifying. A column-major input is unlikely to alter the best-performing combination of these parameters.
